### ISSUE TITLE
Bad LCP and SI on text-dominant site

### ISSUE THREAD

**Issue by brendankenny** [GoogleChrome/lighthouse#11014]
We've discussed this case in context with some other LCP issues, but filing here for external tracking and to see if there is anything specific to a text-only LCP that's useful for testing.

#### Provide the steps to reproduce
1. Run LH on https://www.danshappir.com/

Notice that the LCP is text using a system font, so doesn't require anything more than the content and styling, both of which are in the main document.
 
#### What is the current behavior?
Using default (simulated throttling), LCP is slower than FCP, and SI is pretty hurt.

#### What is the expected behavior?
LCP and FCP should be equal, and SI should be lower.

My CLI result:
https://googlechrome.github.io/lighthouse/viewer/?gist=393fb1db0e3075eb5a41a0910abf0a5a

WPT result on a moto G4/3GFast
https://www.webpagetest.org/lighthouse.php?test=200624_GB_32a10143f45d3212c1b771d9b76cfa0a&run=1

cc @DanShappir

**Comment by patrickhulce** [comment]
Unfortunately there's not really anything that we can do from the simulation side of things that doesn't break most other sites. 

The root issue here is that there is a lot of script execution very early such that the first paint event is embedded in the same task as lots of other script work. 

![image](https://user-images.githubusercontent.com/2301202/85758954-64493f80-b6d6-11ea-94a2-db9c9406242c.png)

In ~90% of cases this means that the script work inside that same task is what enabled the paint. Here though the heavy use of link headers for preloading and async execution of those scripts while the main document was finishing means that Chrome wasn't able to paint the content earlier (Chrome cannot stop in the middle of a task). You can *only* observe this behavior once the network is stretched out.

**Comment by DanShappir** [comment]
Hi Patrick and Brendan,

Thank you for taking the time to review this. If I understand the response correctly, the cause of the large difference between the LCP I'm seeing for the same page in local LH (Chrome Canary dev tools) and PSI has to do with network variations, correct? This is despite the fact that both use simulated network throttling to alleviate network variance impact.

Also, looking at the screenshots produced for a PSI run, I don't see any visual change in the page after the initial render:
<img width="910" alt="Screen Shot 2020-06-26 at 12 37 42" src="https://user-images.githubusercontent.com/8604327/85843537-f66e4800-b7a9-11ea-9bc9-d14bbad8e8bf.png">
Therefore, I don't understand why there is such a large gap between FCP and LCP in this case. Either FCP is too early or LCP is too late.

And, again, this is happening in PSI, but not in local LH in Chrome Canary.

Am I missing something?

**Comment by patrickhulce** [comment]
> the cause of the large difference between the LCP I'm seeing for the same page in local LH (Chrome Canary dev tools) and PSI has to do with network variations, correct?

Not really unless you have some *extreme* network variability. The cause has to do with the fact that script execution when the network is fast occurs in the same tasks as these first paints and inherent implications that has for simulation. This is not variable from any of my observations. The execution pattern always looks like this on fast networks.

The difference between LCP and FCP here is the fact that simulation for FCP more aggressively removes "potentially render-blocking" scripts from its graphs than LCP does.

> And, again, this is happening in PSI, but not in local LH in Chrome Canary.

This is unexpected to me. This problem exists in simulated throttling in all environments and I see very similar results to PSI in local Chrome Canary unless I uncheck "Simulated throttling".

**Comment by andy-feldman** [comment]
I'm going to have to agree with @brendankenny on this one -- definitely looks like extreme variations in the LCP metric and non-adherence to the core principle of what LCP is meant to measure.  It's meant to measure when the user can visibly see the largest visual element, but from my experience on Lighthouse 6.0 and checking on various sites, LCP is often logged extremely late in the trace well after all above-the-fold painting had already completed.  Even PageSpeed Insights is showing a wildly varying difference between field data and lab data for LCP.

Considering that Google had already adopted this new initiative of "Core Web Vitals" including LCP into Search Console globally and is affecting countless websites, I'm not sure if stating that LCP does not "aggressively remove" potentially render-blocking scripts is a responsibly adequate culmination for a likely serious issue where LCP actually occurs long before LCP is guesstimated to have occurred.  In my opinion, Lighthouse 6.0 is still in beta stages with regards to LCP and should not have been rolled out yet with Google.

**Comment by DanShappir** [comment]
I tested again with PSI and with LH in Chrome Canary 86.0.4185.0 (Official Build) canary (64-bit)
Page: https://www.danshappir.com/
With PSI I consistently get a difference of ~1 second between FCP and LCP for this page - see my previous comment for an example.
With LH in Canary, with Simulated Throttling enabled, I consistently get a difference of up to 0.1 second, for example:
<img width="418" alt="Screen Shot 2020-06-28 at 9 56 22" src="https://user-images.githubusercontent.com/8604327/85940614-17f43e80-b926-11ea-943a-b03fa568b6a1.png">
I don't understand the difference between these results

**Comment by patrickhulce** [comment]
If you've come here to complain about Google's decisions @andy-feldman I'm afraid there's not much I can contribute to that conversation.

If, however, you'd like to have a good faith discussion about the difficulty and edge cases of simulated throttling and how they can be addressed, then I'll happily continue :)

> With LH in Canary, with Simulated Throttling enabled, I consistently get a difference of up to 0.1 second, for example:

Fascinating! From PSI, my local, and @brendankenny 's local results we tend to fall into the same execution pattern discussed in https://github.com/GoogleChrome/lighthouse/issues/11014#issuecomment-649674585 . Something about your local environment must have enough latency to fall into the slow network execution pattern. I apologize for my narrow interpretation of your use of "variability" before. It seems environmental variability is in fact to blame in the difference you're seeing with PSI even if it doesn't explain the differences I was seeing. 

In terms of what to do about it, there is nothing that can really be done about the specific paint trace event issue lying at the root of the difference. It's not possible for a human looking at the trace to know if the work inside that task was or wasn't required for LCP to occur or not, and so it will not be possible for simulation either. 

There are some additional heuristics that could be applied to dampen the magnitude of the overestimation in this specific case such as removing images from the graph completely if the last largest contentful element wasn't marked as an image, but it will not ultimately make LCP equivalent to FCP in this case.

> It's meant to measure when the user can visibly see the largest visual element, but from my experience on Lighthouse 6.0 and checking on various sites, LCP is often logged extremely late in the trace well after all above-the-fold painting had already completed.

That's what it's *meant* to capture, yes, but that's not what the metric definition *actually* captures. It approximates when the user experiences these changes based on when contentful elements were loaded, depending on content type, and there are plenty of edge cases where that doesn't align with visual page changes. Chrome speed metrics team is already looking into some changes to improve those edge cases surrounding removal of elements, but the context is important to understand that, simulation or not, LCP is not a magical metric that perfectly captures when the user can visibly see the largest visual content. Like all heuristic-based metrics, there will be some imperfect cases. 

The larger question when it comes to the imperfect cases is if the metric still incentivizes the correct behavior, even if it's not capturing the exact intended moment. I can fairly confidently say that the answer here is still yes. It's unfortunate that LCP is significantly overstated but it's also true that TBT is dramatically understated.  The cause behind both is the amount of script execution. The steps to address the overstated LCP would indeed improve the performance of the page on the target device with TBT, so even if the metric in this case isn't capturing the moment we hope it would, it's still incentivizing the right corrective action on the page as a whole.

**Comment by DanShappir** [comment]
Thank you @patrickhulce for the detailed response and explanation.
To verify that I properly understand your statements, here is my brief summary of them:
The difference in results are caused by the application of the simulated throttling when applied to the measured metrics.
Due to actual network differences, the application of the simulated throttling on my computer excluded script execution time, while in PSI and on your computer, it did include them.
As a result, PSI placed the LCP after the script blocked finished, while FCP in all cases was placed before the script started.
This is why PSI and your measurements show the 1 second difference (which is part of the blocking time), while my measurements do not.
Is my understanding correct?
Thanks

**Comment by patrickhulce** [comment]
Yes, everything you just said from the way I understood it is correct. 👍 

Basically

```
Machines with the gap in FCP/LCP

          |------------------------█████|
start of trace                                script    LCP

Machines without the gap FCP/LCP

          |------------------------|---█████ 
start of trace                            LCP     script

**Comment by DanShappir** [comment]
Got it - thanks again.

**Comment by DanShappir** [comment]
As I final point, I will say that it is unfortunate that the simulated throttling mechanism, which is also intended to make the measurements "more deterministic", creates such a discrepancy in this particular case. And that it results in an incorrect LCP value being shown. I hope this doesn't impact too many Wix sites. (BTW we are currently engaged in an effort to dramatically reduce amount of JS we download and run - hopefully this will eliminate, or at least mitigate, this issue.)

**Comment by teclone** [comment]
From what I found out so far, Largest contentful paint almost like 1s ahead of the first contentful paint on lower-end devices, even when there are no scripting elements. On desktop, it is like 0.3s ahead of the first contentful paint.

I have this [s3 bucket address](https://s3.eu-central-1.amazonaws.com/dev.clacified.com.website/en/index.html), you can take a look.

There is totally no scripting on this page because the scripting links points to a wrong location, this is intentional. The largest contentful element on the page is an image (both mobile and desktop). This image is also preloaded using the link element on the head.

The other link with scripting enabled is at [here](https://dev.clacified.com/en). You will find out that the results for FCP and LCP are the same even though one is scripted and the other has no scripting.

I will also point out another thing. The scripts on the second page are loaded using preload. The scripts itself is added at the end body tag, the intention is to avoid any render-blocking time. It is a gatsby static generated site as well, just to provide more context.

Despite the way the scripts are arranged (appended at the body end tag), the total blocking time is huge. I then wonder how the script manages to block the UI. I can understand the time to interactive value, because the page will not be interactive during the react hydration process, maybe I misunderstood this one really.

I will be very glad to hear your expert reviews.

In conclusion, I am saying that the difference between FCP & LCP is not really dependent on the scripting as those two URLs points out, Maybe this might need to be addressed further.

Cheers

**Comment by patrickhulce** [comment]
@teclone you are right in your conclusion that large script executions were not the reason for the difference in the pages you linked, but the true cause has nothing to do with the issue discussed here. 

This page has a legitimate difference between FCP and LCP. Something removes all of the content and repaints it.

![image](https://user-images.githubusercontent.com/2301202/87249218-50fed980-c423-11ea-8451-e6ccb594d489.png)

Lighthouse is correctly pointing out the problem and this is unrelated what's discussed in this issue.

> Despite the way the scripts are arranged (appended at the body end tag), the total blocking time is huge. I then wonder how the script manages to block the UI. 

There's a misunderstanding with Total Blocking Time. The way scripts are arranged has no impact on TBT, the *duration* of all your script work is what impacts TBT. Please see [the documentation](https://web.dev/tbt/).

**Comment by teclone** [comment]
wow @patrickhulce , thanks for pointing this out, I have been noticing this sort of thing but could not understand this. Please what could be the reason for this? Any info would be highly appreciated

**Comment by patrickhulce** [comment]
Since you mentioned gatsby I'm guessing there's an issue with the hydration process, but I can't say for certain. I don't know what was done to the "non-scripting" version.

**Comment by teclone** [comment]
The non scripting version, the script src attributes points to a wrong file location, so there is no javascript executed there. I know this will have to do with the hydration process. I should use a pure component for the top level page to prevent this from happening?

Also, could it be related to the use of render callbacks on react components? I am not an expert on this, I am just wondering, I have spent many days trying to figure out what could be the cause of this

**Comment by patrickhulce** [comment]
It's not really our role to figure out what's going on with specific tech stacks 😅 Maybe you'll have better luck in [the gatsby community](https://www.gatsbyjs.org/contributing/community/#where-to-get-support)?

**Comment by teclone** [comment]
> It's not really our role to figure out what's going on with specific tech stacks  Maybe you'll have better luck in [the gatsby community](https://www.gatsbyjs.org/contributing/community/#where-to-get-support)?

I have create a ticket on gatsby regarding this. thanks @patrickhulce

**Comment by teclone** [comment]
> @teclone you are right in your conclusion that large script executions were not the reason for the difference in the pages you linked, but the true cause has nothing to do with the issue discussed here.
> 
> This page has a legitimate difference between FCP and LCP. Something removes all of the content and repaints it.
> 
> ![image](https://user-images.githubusercontent.com/2301202/87249218-50fed980-c423-11ea-8451-e6ccb594d489.png)
> 
> Lighthouse is correctly pointing out the problem and this is unrelated what's discussed in this issue.

I have been investigating this @patrickhulce, and it turns out that nothing is removing the page contents and rendering it again. Please can you check this on other websites or check the [unscripted version of the page](https://s3.eu-central-1.amazonaws.com/dev.clacified.com.website/en/index.html).

This is what I think is happening, based on my findings, once the browser fires domContentLoaded, somehow, chrome rerenders the page to calculate any layout shift. especially images. You can create a basic HTML page, with an image in it, and check for yourself. no scripts.

This is just a guess based on what I saw. You can find check again without your expert understanding. Looking forward to your feedback.

**Comment by patrickhulce** [comment]
Removing page contents was just one plausible explanation for what is observed in the trace I of course don't know how specifically this page was built :) The fact of the matter is that *something* the page is doing causes an artificial delay in reaching its final LCP.

> then lighthouse scripts should record the largest contentful paint before or during the domContentLoaded event, and not after onLoad event.

Lighthouse records LCP as whatever the browser reported LCP to be. When there is a legitimate difference between FCP and LCP in the browser's eyes, then there will be in Lighthouse as well and such differences are *expected*, not a bug that needs investigation. 

What was discussed earlier in this issue was a very specific site that has identical FCP/LCP in the browser, but not in Lighthouse. *That* was worthy of investigation. Unfortunately, this page example is not the same situation. Lighthouse will not do anything differently for this page.

**Comment by andy-feldman** [comment]
Users who find that Chrome has been tricked into reporting LCP long after painting had already completed may wish to consider filing a bug report for Chromium: https://crbug.com/wizard

**Comment by patrickhulce** [comment]
There's no specific action for Lighthouse to take at the moment here. If you find bugs with the LCP metric itself, they can be reported in the link referenced in https://github.com/GoogleChrome/lighthouse/issues/11014#issuecomment-658992362
